src/webview/ConnectionPanel.ts

```diff
@@ -0,0 +1,322 @@
+      type: 'Connection.InitializeView',
+      connection: oldConnection,
+    });
+  }
+
+  public static revive(panel: vscode.WebviewPanel, extensionUri: vscode.Uri, mode: PanelMode) {
+    ConnectionPanel.currentPanel = new ConnectionPanel(panel, extensionUri, mode);
+  }
+
+  private constructor(panel: vscode.WebviewPanel, extensionUri: vscode.Uri, mode: PanelMode) {
+    this._panel = panel;
+    this._extensionUri = extensionUri;
+
+    // Set the webview's initial html content
+    this._update(mode);
+
+    // Listen for when the panel is disposed
+    // This happens when the user closes the panel or when the panel is closed programmatically
+    this._panel.onDidDispose(() => this.dispose(), null, this._disposables);
+
+    // Update the content based on view changes
+    this._panel.onDidChangeViewState(
+      e => {
+        if (this._panel.visible) {
+          this._update(mode);
+        }
+      },
+      null,
+      this._disposables,
+    );
+
+    // Handle messages from the webview
+    this._panel.webview.onDidReceiveMessage(
+      message => {
+        switch (message.type) {
+          case 'Connection.cancel': {
+            this.dispose();
+            break;
+          }
+          case 'Connection.testConnection': {
+            const panel = vscode.window.createWebviewPanel(
+              'catCoding',
+              'Cat Coding',
+              ViewColumn.One,
+              {},
+            );
+
+            panel.webview.html = `<!DOCTYPE html>
+              <html lang="en">
+              <head>
+                  <meta charset="UTF-8">
+                  <meta name="viewport" content="width=device-width, initial-scale=1.0">
+                  <title>Cat Coding</title>
+              </head>
+              <body>
+                  <img src="https://media.giphy.com/media/JIX9t2j0ZTN9S/giphy.gif" width="300" />
+              </body>
+              </html>`;
+
+            break;
+          }
+          case 'Connection.add': {
+            this.addConnection(message.name, message.url, message.context);
+            break;
+          }
+          case 'Connection.edit': {
+            this.editConnection(
+              message.name,
+              message.url,
+              message.context,
+              message.oldConnectionName,
+            );
+            break;
+          }
+        }
+      },
+      null,
+      this._disposables,
+    );
+  }
+
+  public dispose() {
+    ConnectionPanel.currentPanel = undefined;
+
+    // Clean up our resources
+    this._panel.dispose();
+
+    while (this._disposables.length) {
+      const x = this._disposables.pop();
+      if (x) {
+        x.dispose();
+      }
+    }
+  }
+
+  public addConnection(name: string, url: string, context: string) {
+    if (name != '') {
+      let aConnection: Connection = {
+        name: name,
+        url: new URL(url),
+        contexts: {},
+      };
+
+      ConnectionManager.connectionManager.upsertConnection(aConnection);
```

Comment:
I have some thoughts about how the ConnectionManager singleton's external interface is presented - let's table that for later, though.

Reply:
The connection manager will be exposed as a getter instead of a public static member.